importccl: WIP: DO NOT MERGE

miretskiy opened the pull request:
Make importing data into cockroach from external http servers
resilient to connection interruption.

If the download is interrupted, and if the external server
supports "Accept-Ranges: bytes" (and most servers do support that),
we attempt to request the next "Content-Range" from the server.

Release note (enterprise change): more resilient http import

Commit:
importccl: WIP: DO NOT MERGE

Make importing data into cockroach from external http servers
resilient to connection interruption.

If the download is interrupted, and if the external server
supports "Accept-Ranges: bytes" (and most servers do support that),
we attempt to request the next "Content-Range" from the server.

Release note (enterprise change): more resilient http import

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43385)
<!-- Reviewable:end -->

dt committed:
importccl: limit concurrent IMPORT workers

This change limits the number of IMPORT worker processors that can execute concurrently.
This is motivated by observations of a customer cluster that is being sent hundreds of
IMPORT jobs and attempting to execute all of them simultainiously and then seeing flakes
from the network connections and transports, potentially because the extreme concurrency
means individual workers are left blocked or unscheduled and fail to read from their
connection for long periods of time, or more generally that the over-taxed node fails to
maintain connections for other reasons.

While some concurency likely does improve utilization, excessive concurrency is likely
harmful beyond the above discussed network flakes -- due to cache evictions thoughout
the pipeline, context switches, etc so a reasonable limit here is unlikely to make
concurrent IMPORTs slower and may in fact make them faster.

The ideal value for this setting is almost certainly different for different clusters
and differnt hardware or deployments, so ideally this would at least be an easy to
tune cluster setting, however the plumbing involved for that is significant for a
change in a patch release. Thus a env var is used here (for 19.2).

Release note (performance improvement): add a limit on the number of concurrent IMPORT workers per node.
